Added writeGltf stage

leerichard42 opened the pull request:
Branched the writeGltf branch off loadShaders, because I used the same async package.

pjcozzi commented:
Nice progress, @leerichard42!

pjcozzi commented on lib/writeGltf.js:
```diff
@@ -0,0 +1,73 @@
+                    throw err;
+                }
+            }
+            else if (callback) {
```
We shouldn't call `callback` until we're inside the callback passed to `writeFile`, correct?

leerichard42 replied:
Not too sure what this means - do you mean the `asyncCallback` passed to the `writeFile` in `writeSource`? `callback` only gets run after all the sources have been written out.

pjcozzi replied:
The `callback` passed to `writeGltf` should not get called until the glTF is written to disk, so it shouldn't get called until in the callback passed to `writeFile` below, correct?

leerichard42 replied:
I see! Yes, I will move `callback` to be called through `writeFile`.

pjcozzi commented on lib/writeGltf.js:
```diff
@@ -0,0 +1,73 @@
+        var sourceDataUri = new dataUri();
+        sourceDataUri.format('.txt', source);
+        shader.uri = sourceDataUri.content;
+        asyncCallback();
```
Here and below, use `process.nextTick`.

pjcozzi commented on lib/writeGltf.js:
```diff
@@ -0,0 +1,73 @@
+        asyncCallback();
+    }
+    else {
+        var fileName = id + '.glsl';
```
Given that we don't want to overwritten the input asset (I imagine in most cases) perhaps we should change the `-o` options in gltf-pipeline.js to be a directory (we can create it if it does not exist)?  What do you think?

pjcozzi replied:
Or have a separate command-line option for outputting to a directory.

leerichard42 replied:
I like the idea of having a separate option to output to a directory, because I imagine that different users would or wouldn't want a new directory depending on their project.

pjcozzi commented on specs/lib/writeGltfSpec.js:
```diff
@@ -0,0 +1,92 @@
+        };
+
+        writeGltf(gltf, './specs/errorFilePath/output.gltf', false, function(err) {
+            expect(function() {
```
Won't just `expect(err).toBeDefined();` be OK?

pjcozzi commented:
Once again, nice job here!  These are some great foundations.

leerichard42 commented:
Updated! Added an option to create an 'output' directory in the destination folder for now.

pjcozzi commented:
@leerichard42  I'll merge master in and fix the conflicts.

pjcozzi commented:
Merged in master.  All the tests pass on CI, but 3 tests fail locally.  Any ideas?

```

> gltf-pipeline@0.0.0 jasmine c:\Users\pjcozzi\source\gltf-pipeline
> jasmine

Started
.......................FF......................................F.

Failures:
1) loadGltfUris loads an embedded shader
  Message:
    Expected 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    ' to equal 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    '.
  Stack:
    Error: Expected 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    ' to equal 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    '.
        at c:\Users\pjcozzi\source\gltf-pipeline\specs\lib\loadGltfUrisSpec.js:49:66
        at c:\Users\pjcozzi\source\gltf-pipeline\lib\loadGltfUris.js:37:21
        at process._tickCallback (node.js:355:11)

2) loadGltfUris loads an external and an embedded shader
  Message:
    Expected 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    ' to equal 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    '.
  Stack:
    Error: Expected 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    ' to equal 'precision highp float;
    varying vec3 v_normal;
    varying vec2 v_texcoord0;
    uniform sampler2D u_diffuse;
    uniform vec4 u_specular;
    uniform float u_shininess;
    void main(void) {
    vec3 normal = normalize(v_normal);
    vec4 color = vec4(0., 0., 0., 0.);
    vec4 diffuse = vec4(0., 0., 0., 1.);
    vec4 specular;
    diffuse = texture2D(u_diffuse, v_texcoord0);
    specular = u_specular;
    diffuse.xyz *= max(dot(normal,vec3(0.,0.,1.)), 0.);
    color.xyz += diffuse.xyz;
    color = vec4(color.rgb * diffuse.a, diffuse.a);
    gl_FragColor = color;
    }
    '.
        at c:\Users\pjcozzi\source\gltf-pipeline\specs\lib\loadGltfUrisSpec.js:70:74
        at c:\Users\pjcozzi\source\gltf-pipeline\lib\loadGltfUris.js:37:21
        at process._tickCallback (node.js:355:11)

3) writeGltf writes an embedded shader
  Message:
    Expected 'data:text/plain;base64,cHJlY2lzaW9uIGhpZ2hwIGZsb2F0Ow0KdmFyeWluZyB2ZWMzIHZfbm9ybWFsOw0KdmFyeWluZyB2ZWMyIHZfdGV4Y29vcmQwOw0KdW5pZm9ybSBzYW1wbGVyMkQgdV9kaWZmdXNlOw0KdW5pZm9ybSB2ZWM0IHVfc3BlY3VsYXI7DQp1bmlmb3JtIGZsb2F0IHVfc2hpbmluZXNzOw0Kdm9pZCBtYWluKHZvaWQpIHsNCnZlYzMgbm9ybWFsID0gbm9ybWFsaXplKHZfbm9ybWFsKTsNCnZlYzQgY29sb3IgPSB2ZWM0KDAuLCAwLiwgMC4sIDAuKTsNCnZlYzQgZGlmZnVzZSA9IHZlYzQoMC4sIDAuLCAwLiwgMS4pOw0KdmVjNCBzcGVjdWxhcjsNCmRpZmZ1c2UgPSB0ZXh0dXJlMkQodV9kaWZmdXNlLCB2X3RleGNvb3JkMCk7DQpzcGVjdWxhciA9IHVfc3BlY3VsYXI7DQpkaWZmdXNlLnh5eiAqPSBtYXgoZG90KG5vcm1hbCx2ZWMzKDAuLDAuLDEuKSksIDAuKTsNCmNvbG9yLnh5eiArPSBkaWZmdXNlLnh5ejsNCmNvbG9yID0gdmVjNChjb2xvci5yZ2IgKiBkaWZmdXNlLmEsIGRpZmZ1c2UuYSk7DQpnbF9GcmFnQ29sb3IgPSBjb2xvcjsNCn0NCg==' to equal 'data:text/plain;base64,cHJlY2lzaW9uIGhpZ2hwIGZsb2F0Owp2YXJ5aW5nIHZlYzMgdl9ub3JtYWw7CnZhcnlpbmcgdmVjMiB2X3RleGNvb3JkMDsKdW5pZm9ybSBzYW1wbGVyMkQgdV9kaWZmdXNlOwp1bmlmb3JtIHZlYzQgdV9zcGVjdWxhcjsKdW5pZm9ybSBmbG9hdCB1X3NoaW5pbmVzczsKdm9pZCBtYWluKHZvaWQpIHsKdmVjMyBub3JtYWwgPSBub3JtYWxpemUodl9ub3JtYWwpOwp2ZWM0IGNvbG9yID0gdmVjNCgwLiwgMC4sIDAuLCAwLik7CnZlYzQgZGlmZnVzZSA9IHZlYzQoMC4sIDAuLCAwLiwgMS4pOwp2ZWM0IHNwZWN1bGFyOwpkaWZmdXNlID0gdGV4dHVyZTJEKHVfZGlmZnVzZSwgdl90ZXhjb29yZDApOwpzcGVjdWxhciA9IHVfc3BlY3VsYXI7CmRpZmZ1c2UueHl6ICo9IG1heChkb3Qobm9ybWFsLHZlYzMoMC4sMC4sMS4pKSwgMC4pOwpjb2xvci54eXogKz0gZGlmZnVzZS54eXo7CmNvbG9yID0gdmVjNChjb2xvci5yZ2IgKiBkaWZmdXNlLmEsIGRpZmZ1c2UuYSk7CmdsX0ZyYWdDb2xvciA9IGNvbG9yOwp9Cg=='.
  Stack:
    Error: Expected 'data:text/plain;base64,cHJlY2lzaW9uIGhpZ2hwIGZsb2F0Ow0KdmFyeWluZyB2ZWMzIHZfbm9ybWFsOw0KdmFyeWluZyB2ZWMyIHZfdGV4Y29vcmQwOw0KdW5pZm9ybSBzYW1wbGVyMkQgdV9kaWZmdXNlOw0KdW5pZm9ybSB2ZWM0IHVfc3BlY3VsYXI7DQp1bmlmb3JtIGZsb2F0IHVfc2hpbmluZXNzOw0Kdm9pZCBtYWluKHZvaWQpIHsNCnZlYzMgbm9ybWFsID0gbm9ybWFsaXplKHZfbm9ybWFsKTsNCnZlYzQgY29sb3IgPSB2ZWM0KDAuLCAwLiwgMC4sIDAuKTsNCnZlYzQgZGlmZnVzZSA9IHZlYzQoMC4sIDAuLCAwLiwgMS4pOw0KdmVjNCBzcGVjdWxhcjsNCmRpZmZ1c2UgPSB0ZXh0dXJlMkQodV9kaWZmdXNlLCB2X3RleGNvb3JkMCk7DQpzcGVjdWxhciA9IHVfc3BlY3VsYXI7DQpkaWZmdXNlLnh5eiAqPSBtYXgoZG90KG5vcm1hbCx2ZWMzKDAuLDAuLDEuKSksIDAuKTsNCmNvbG9yLnh5eiArPSBkaWZmdXNlLnh5ejsNCmNvbG9yID0gdmVjNChjb2xvci5yZ2IgKiBkaWZmdXNlLmEsIGRpZmZ1c2UuYSk7DQpnbF9GcmFnQ29sb3IgPSBjb2xvcjsNCn0NCg==' to equal 'data:text/plain;base64,cHJlY2lzaW9uIGhpZ2hwIGZsb2F0Owp2YXJ5aW5nIHZlYzMgdl9ub3JtYWw7CnZhcnlpbmcgdmVjMiB2X3RleGNvb3JkMDsKdW5pZm9ybSBzYW1wbGVyMkQgdV9kaWZmdXNlOwp1bmlmb3JtIHZlYzQgdV9zcGVjdWxhcjsKdW5pZm9ybSBmbG9hdCB1X3NoaW5pbmVzczsKdm9pZCBtYWluKHZvaWQpIHsKdmVjMyBub3JtYWwgPSBub3JtYWxpemUodl9ub3JtYWwpOwp2ZWM0IGNvbG9yID0gdmVjNCgwLiwgMC4sIDAuLCAwLik7CnZlYzQgZGlmZnVzZSA9IHZlYzQoMC4sIDAuLCAwLiwgMS4pOwp2ZWM0IHNwZWN1bGFyOwpkaWZmdXNlID0gdGV4dHVyZTJEKHVfZGlmZnVzZSwgdl90ZXhjb29yZDApOwpzcGVjdWxhciA9IHVfc3BlY3VsYXI7CmRpZmZ1c2UueHl6ICo9IG1heChkb3Qobm9ybWFsLHZlYzMoMC4sMC4sMS4pKSwgMC4pOwpjb2xvci54eXogKz0gZGlmZnVzZS54eXo7CmNvbG9yID0gdmVjNChjb2xvci5yZ2IgKiBkaWZmdXNlLmEsIGRpZmZ1c2UuYSk7CmdsX0ZyYWdDb2xvciA9IGNvbG9yOwp9Cg=='.
        at Object.<anonymous> (c:\Users\pjcozzi\source\gltf-pipeline\specs\lib\writeGltfSpec.js:64:59)

65 specs, 3 failures
Finished in 0.157 seconds

```

pjcozzi commented on lib/writeGltf.js:
```diff
@@ -0,0 +1,85 @@
+        }, function(err) {
+            if (err) {
+                if (callback) {
+                        process.nextTick(function() {
```
I don't think I was clear about this before, if we call a callback from within a callback (like this case here since we are calling `callback` inside the callback passed to `async.each` then we do not need to use `proces.nextTick` since the first callback should have handled it.  I _think_ this is the best practice to keep the code from being too verbose.

pjcozzi replied:
This comment applies throughout.

leerichard42 replied:
I think I see what you mean, but when you say that the first callback should have handled it, do you mean since `asyncCallback` was called using `process.nextTick`, then that also accounts for `callback`? Because I'm not manually calling the callback passed to `async.each`, so I'm not sure how it is being called.

pjcozzi replied:
Exactly.  Basically, all callbacks need to be async, e.g., as if they are called after file or network IO, so a function should never call a callback synchronously; instead, they should use `process.nextTick` so it happens later in Node's runtime loop (it is kinda like faking very fast file IO).  So, in the case of decoding a data URI, we could call the callback synchronously for data URIs, but it would be a bad practice because, from the user's perspective, sometimes it would be called synchronously (data URI) and sometimes not (file url) so the Node.js best practice is to always make it async.

leerichard42 replied:
I see! But in that case, would I still need to use `process.nextTick` within the callbacks for `writeFile` and `readFile`? When you first commented in the previous PR, it was for a callback within `readFile`.

leerichard42 replied:
(In other words, I would only use `process.nextTick` to make the data URI case asynchronous, right?)

pjcozzi replied:
> When you first commented in the previous PR, it was for a callback within readFile.

Sorry, I was wrong there.

> (In other words, I would only use process.nextTick to make the data URI case asynchronous, right?)

Exactly because you want the callback to appear to be async.

pjcozzi commented:
Just those comments.  This is close!

pjcozzi commented:
@leerichard42 let me know when this is ready.  The next step (in a separate PR) is to expand read/writeGltf to include buffers and images (both external urls and data uris).  Once we have that, we'll have everything we need to convert a glTF asset between embedded and external resources.  Pretty cool.

leerichard42 commented:
Updated! I'm not too sure about why those tests failed, but I actually had a similar problem myself locally which I fixed by converting the data to strings before testing for equality. (e.g. `expect(gltf.shaders.CesiumTexturedBoxTest0FS.extras.source.toString()).toEqual(fragmentShaderData.toString());`

pjcozzi commented:
I'll write an issue for the test failures, and I'll look at it later.

Let's merge this so you can move on to buffers and images as mentioned above, https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/32#issuecomment-183147014.  Let me know if you have any questions.

You're starting to build some exciting foundations!
